OAuth documentation overhaul (#2300)
clerkinc/clerk-docs: docs/oauth/oauth-scoped-access.mdx

> [!NOTE]
> If you would like to know more about the terminology around OAuth, check out our [OAuth terminology guide](/docs/oauth/oauth-overview#key-terminology).

This guide provides step-by-step instructions on how to build OAuth scoped access into a Clerk application.

## Before you start

- You need to [create a Clerk account](https://dashboard.clerk.com/sign-up) or [sign into the Clerk Dashboard](https://dashboard.clerk.com/).
- You need to create a Clerk Application in the [Clerk Dashboard](https://dashboard.clerk.com/). For more information, see the [setup guide](/docs/quickstarts/setup-clerk).

Let's begin by creating an OAuth application in the Clerk Dashboard.

<Steps>
  ### Create a Clerk OAuth application

  1. In the Clerk Dashboard, navigate to the [**OAuth Applications**](https://dashboard.clerk.com/last-active?path=user-authentication/oauth-applications) page.

  1. Select the **Add OAuth application** button. A modal will open.

  1. Complete the following fields:

     - `Name` - Helps you identify your application.

>      - `Scopes` - The scopes that you would like to leverage. To find out more about the scopes currently supported by Clerk, check out the [OAuth reference documentation](/docs/oauth/how-clerk-implements-oauth#scopes).

Reviewer: There are a few more fields when creating an oauth application.

You have to enable staff mode to see the "Consent Screen" toggle

![image](https://github.com/user-attachments/assets/b249ed5d-fa67-4195-87bf-b0cc825cb1c7)

Author: This is all addressed in the How Clerk implements OAuth page

>      ![A screenshot of a new OAuth application being created in the Clerk dashboard](/docs/images/oauth/dashboard-create-oauth-application.png)

Reviewer: Could be an issue on my end, but I don't see this screenshot when looking at the preview: https://clerk.com/docs/pr/ss-docs-10474/oauth/oauth-single-sign-on#create-a-clerk-o-auth-application

  1. In the Clerk Dashboard, navigate to the [**OAuth Applications**](https://dashboard.clerk.com/last-active?path=user-authentication/oauth-applications) page.

  1. Select the **Add OAuth application** button. A modal will open.

  1. Complete the following fields:

     - `Name` - Helps you identify your application.
     - `Scopes` - The scopes that you would like to leverage. To find out more about the scopes currently supported by Clerk, check out the [OAuth reference documentation](/docs/oauth/how-clerk-implements-oauth#scopes).

     ![A screenshot of a new OAuth application being created in the Clerk dashboard](/docs/images/oauth/dashboard-create-oauth-application.png)

  1. Select **Add**. A modal will open with your **Client Secret**.

     > [!WARNING]
     > For security reasons, Clerk does not store your Client Secret and cannot show it to you again, so make sure to copy the secret and store it someplace secure, like an environment variable.

  1. You'll be redirected to your OAuth application's settings page. Under **Application credentials**, you can find your **Client ID**, which will be necessary for to complete the OAuth flow.

>   1. In the **Redirect URIs** field, add the redirect URI provided by your client. This is the URL that Clerk will redirect to with an authorization code after the user has authenticated and consented to the access request. The example below uses [http://localhost:3000/oauth\_callback](http://localhost:3000/oauth_callback).

Reviewer 2: ```suggestion
  1. In the **Redirect URIs** field, add the redirect URI provided by your client. This is the URL that Clerk will redirect to with an authorization code after the user has authenticated and consented to the access request. For this example, we will use [http://localhost:3000/oauth\_callback](http://localhost:3000/oauth_callback).
```

     ![A screenshot of the redirect urls, client id, and client secret within an OAuth application in the Clerk dashboard](/docs/images/oauth/dashboard-oauth-client-details.png)

  The Clerk setup is now complete. If you'd like to know more about other OAuth application configuration settings, refer to the [dedicated reference documentation](/docs/oauth/how-clerk-implements-oauth).

  ### Test the OAuth flow

>   In order to test the OAuth flow, you can use the following [OAuth test client](https://github.com/clerk/test-oauth-client) project. It contains a minimal Express.js server that can run through an OAuth flow.

Reviewer 2: ```suggestion
  If you are aiming to integrate your Clerk application with a third party app that acts as an OAuth client, you can reference their documentation on how to integrate OAuth connections in order to test your OAuth flow. It will likely be necessary to provide the `client_id`, `client_secret`, and `redirect_uri` from the OAuth application we created above.
  
  If you don't have a specific, existing OAuth client in mind, you can use the following [OAuth test client](https://github.com/clerk/test-oauth-client) project. It contains a minimal Express.js server that can run through an OAuth flow. We'll walk through how to use the test client below.
```

  > [!NOTE]
  > The consent screen uses the scopes passed in the authorization request to inform the user exactly what they’re granting access to, and to whom.
  >
  > By default, the consent screen is shown for all newly created OAuth apps, but this can be toggled in each app’s settings [on the Clerk Dashboard](https://dashboard.clerk.com/last-active?path=oauth-applications), although not recommended.
  >
  > For more details on how our Clerk OAuth consent screen works, see the [OAuth consent screen reference docs](/docs/oauth/how-clerk-implements-oauth#consent-screen-management).

  Once you have accepted, you will be redirected to the OAuth callback route, the `redirect_uri` specified earlier. This process will exchange the authorization code for an access token, and render the access token and refresh token as a JSON response, similar to this:

  ```json
  {
    "access_token": "xxx",
    "expires_in": 7200,
    "id_token": "xxx",
    "refresh_token": "xxx",
    "scope": "openid email profile",
    "token_type": "bearer"
  }
  ```
</Steps>

> ## Accepting and verifying an OAuth Token with Clerk SDKs

Reviewer 2: I do still think we should bounce this out to a separate page, then this section at the bottom can link out to that page with something like "with the OAuth flow complete, the client now has a working access token. The next step would normally be for the client to use that token to make an authenticated request to your app's API endpoints. To learn more about how to accept and verify OAuth access tokens using Clerk, see our guide on verifying oauth access tokens (link)"

If you are building an application that uses Clerk and would like to incorporate OAuth, you will want to ensure that, after the client gets an OAuth access token, they can use it to make authenticated requests into your app (the _resource service_) using the token.

Clerk’s SDKs support this through the `acceptsToken` option that can be used in Clerk's route protection functions, such as [`auth()`](/docs/references/nextjs/auth), [`auth.protect()`](/docs/references/nextjs/auth#auth-protect) and [`authenticateRequest()`](/docs/references/backend/authenticate-request).

{/* TODO: Add links to different guides */}

For detailed examples and best practices on accepting and verifying OAuth tokens with Clerk SDKs, see the following guides:

- [Accepting and verifying an OAuth Token with Next.js]().
- [Accepting and verifying an OAuth Token with React Router]().

> - [Accepting and verifying an OAuth Token with Tanstack React Start]().

Author: Okay so maybe this whole things is its own page under Guides as Verifying OAuth access tokens? @jescalan

export default clerkMiddleware(async (auth, req) => {
  if (isOAuthAccessible(req)) await auth.protect({ token: 'oauth_token' })
  if (isUserAccessible(req)) await auth.protect({ token: 'session_token' })
  if (isAccessibleToAnyValidToken(req)) await auth.protect()
})
```

### Example: Node.js with the JS Backend SDK (Hono)

If you aren’t using Next.js, here’s an example of how a request could be verified using [the JS Backend SDK](https://clerk.com/docs/references/backend/overview), with a Node.js framework like [Hono](https://hono.dev/):

```tsx
import { Hono } from 'hono'
import { clerkClient } from '@clerk/backend'

const app = new Hono()

app.use('*', async (ctx, next) => {
  const clerk = createClerkClient({ secretKey: process.env.CLERK_SECRET })

  // can also be "oauth_token" or "session_token"
  const authReq = await clerk.authenticateRequest(req, { acceptsToken: 'any' })
  const authObject = authReq.toAuth()

  if (authObject.tokenType === 'session_token') {
    console.log('this is session token from a user')
  } else if (authObject.tokenType === 'oauth_token') {
    console.log('this is an oauth token')
  }

  await next()
})

export default app

> ```

Author: @jescalan Currently what I'm working on, so yeah will just link things here.

> ```

Author: Here is the PR with that work: https://github.com/clerk/clerk-docs/pull/2304

1. Select **Add**. A modal will open with your **Client Secret**.

   > [!WARNING]
   > For security reasons, Clerk does not store your Client Secret and cannot show it to you again, so make sure to copy the secret and store it someplace secure, like an environment variable.
1. You'll be redirected to your OAuth application's settings page. Under **Application credentials**, you can find your **Client ID**, which will be necessary for to complete the OAuth flow.
1. In the **Redirect URIs** field, add the redirect URI provided by your client. This is the URL that Clerk will redirect to after the user has authenticated.

If you'd like to know more about other OAuth application configuration settings, refer to the [dedicated reference documentation](/docs/oauth/how-clerk-implements-oauth).

## Building an OAuth Client

If you are building an OAuth **client** that you'd like to have the ability to access a Clerk application's users' data via OAuth, it will need to be configured as an OAuth client and correctly implement the OAuth authorization code flow. This process involves:

> - Registering an OAuth client/application.

Author: @jescalan Bit confused by this one - aren't we doing that above? And should this be added as a step below?
